File: zak/core/runtime/executor.py
"""
ZAK AgentExecutor — orchestrates the full agent lifecycle:
  pre_run → ([policy check] → execute) → post_run → audit

The executor is the only place where policy enforcement and audit emission happen.
Agents themselves are never responsible for these cross-cutting concerns.
"""

from __future__ import annotations

import time

from zak.core.audit.events import (
    AgentCompletedEvent,
    AgentFailedEvent,
    AgentStartedEvent,
    AuditEventType,
    PolicyBlockedEvent,
)
from zak.core.audit.logger import AuditLogger
from zak.core.policy.engine import PolicyDecision, PolicyEngine
from zak.core.runtime.agent import AgentContext, AgentResult, BaseAgent
from zak.tenants.context import TenantContext, TenantRegistry


class AgentExecutor:
    """
    Executes an agent within a tenant-scoped context.

    Flow:
      1. Emit agent.started audit event
      2. Call agent.pre_run()
      3. Evaluate pre-execution policy
      4. Call agent.execute()
      5. Call agent.post_run()
      6. Emit agent.completed or agent.failed audit event

    Usage:
        executor = AgentExecutor()
        result = executor.run(my_agent, context)
    """

    def __init__(self) -> None:
        self._policy = PolicyEngine()

    def run(self, agent: BaseAgent, context: AgentContext) -> AgentResult:
        """Execute `agent` within `context`, enforcing policy and emitting audit events."""
        # Block deactivated tenants before any work or audit
        registry = TenantRegistry.get()
        if registry.exists(context.tenant_id):
            tenant_ctx = TenantContext(
                tenant_id=context.tenant_id,
                trace_id=context.trace_id,
                environment=context.environment,
            )
            tenant_ctx.assert_active(registry)

        logger = AuditLogger(
            tenant_id=context.tenant_id,
            agent_id=context.agent_id,
            trace_id=context.trace_id,
        )

        logger.emit(AgentStartedEvent(
            agent_id=context.agent_id,
            tenant_id=context.tenant_id,
            trace_id=context.trace_id,
            payload={"domain": context.dsl.agent.domain.value, "version": context.dsl.agent.version},
        ))

        start = time.monotonic()

        try:
            # Pre-run hook
            agent.pre_run(context)

            # Pre-execution policy evaluation
            policy_check = self._policy.evaluate(
                dsl=context.dsl,
                action="agent_execute",
                environment=context.environment,
            )
            if not policy_check.allowed:
                logger.emit(PolicyBlockedEvent(
                    agent_id=context.agent_id,
                    tenant_id=context.tenant_id,
                    trace_id=context.trace_id,
                    action="agent_execute",
                    reason=policy_check.reason,
                ))
                duration_ms = (time.monotonic() - start) * 1000
                return AgentResult.fail(context, errors=[f"Policy denied: {policy_check.reason}"], duration_ms=duration_ms)

            # Main execution
            result = agent.execute(context)
            result.duration_ms = (time.monotonic() - start) * 1000

            # Post-run hook
            agent.post_run(context, result)

            if result.success:
                logger.emit(AgentCompletedEvent(
                    agent_id=context.agent_id,
                    tenant_id=context.tenant_id,
                    trace_id=context.trace_id,
                    success=True,
                    duration_ms=result.duration_ms,
                ))
            else:
                logger.emit(AgentFailedEvent(
                    agent_id=context.agent_id,
                    tenant_id=context.tenant_id,
                    trace_id=context.trace_id,
                    error="; ".join(result.errors),
                ))

            return result

        except Exception as exc:
            duration_ms = (time.monotonic() - start) * 1000
            logger.emit(AgentFailedEvent(
                agent_id=context.agent_id,
                tenant_id=context.tenant_id,
                trace_id=context.trace_id,
                error=str(exc),
            ))
            return AgentResult.fail(context, errors=[str(exc)], duration_ms=duration_ms)

    def check_action(
        self, context: AgentContext, action: str, environment: str | None = None
    ) -> PolicyDecision:
        """
        Mid-execution policy check for a specific action (callable from within agents).

        Agents should call this before performing sensitive operations.
        """
        logger = AuditLogger(
            tenant_id=context.tenant_id,
            agent_id=context.agent_id,
            trace_id=context.trace_id,
        )
        decision = self._policy.evaluate(
            dsl=context.dsl,
            action=action,
            environment=environment or context.environment,
        )
        if decision.allowed:
            logger.log_raw(AuditEventType.POLICY_ALLOWED, action=action)
        else:
            logger.emit(PolicyBlockedEvent(
                agent_id=context.agent_id,
                tenant_id=context.tenant_id,
                trace_id=context.trace_id,
                action=action,
                reason=decision.reason,
            ))
        return decision

File: adk/node/src/core/runtime/executor.ts
/**
 * ZAK AgentExecutor — orchestrates the full agent lifecycle:
 *   pre_run → ([policy check] → execute) → post_run → audit
 *
 * The executor is the only place where policy enforcement and audit emission happen.
 * Agents themselves are never responsible for these cross-cutting concerns.
 *
 * TypeScript equivalent of zak/core/runtime/executor.py.
 */

import {
  agentCompletedEvent,
  agentFailedEvent,
  agentStartedEvent,
  policyBlockedEvent,
  AuditEventType,
} from "../audit/events.js";
import { AuditLogger } from "../audit/logger.js";
import { PolicyEngine, type PolicyDecision } from "../policy/engine.js";
import type { AgentContext, AgentResult, BaseAgent } from "./agent.js";
import { agentResultFail, getAgentId } from "./agent.js";

// ---------------------------------------------------------------------------
// AgentExecutor
// ---------------------------------------------------------------------------

/**
 * Executes an agent within a tenant-scoped context.
 *
 * Flow:
 *   1. Emit agent.started audit event
 *   2. Call agent.preRun()
 *   3. Evaluate pre-execution policy
 *   4. Call agent.execute()
 *   5. Call agent.postRun()
 *   6. Emit agent.completed or agent.failed audit event
 */
export class AgentExecutor {
  private _policy: PolicyEngine;

  constructor() {
    this._policy = new PolicyEngine();
  }

  /**
   * Execute `agent` within `context`, enforcing policy and emitting audit events.
   */
  async run(agent: BaseAgent, context: AgentContext): Promise<AgentResult> {
    const agentId = getAgentId(context);
    const logger = new AuditLogger(context.tenantId, agentId, context.traceId);

    // 1. Emit agent.started
    logger.emit(
      agentStartedEvent(agentId, context.tenantId, context.traceId, {
        domain: context.dsl.agent.domain,
        version: context.dsl.agent.version,
      })
    );

    const start = performance.now();

    try {
      // 2. Pre-run hook
      agent.preRun(context);

      // 3. Pre-execution policy evaluation
      const policyCheck = this._policy.evaluate(
        context.dsl,
        "agent_execute",
        context.environment
      );

      if (!policyCheck.allowed) {
        logger.emit(
          policyBlockedEvent(
            agentId,
            context.tenantId,
            context.traceId,
            "agent_execute",
            policyCheck.reason
          )
        );
        const durationMs = performance.now() - start;
        return agentResultFail(
          context,
          [`Policy denied: ${policyCheck.reason}`],
          durationMs
        );
      }

      // 4. Main execution (supports sync and async agents)
      let result = agent.execute(context);
      if (result instanceof Promise) {
        result = await result;
      }
      result.durationMs = performance.now() - start;

      // 5. Post-run hook
      agent.postRun(context, result);

      // 6. Emit completion/failure audit event
      if (result.success) {
        logger.emit(
          agentCompletedEvent(
            agentId,
            context.tenantId,
            context.traceId,
            true,
            result.durationMs
          )
        );
      } else {
        logger.emit(
          agentFailedEvent(
            agentId,
            context.tenantId,
            context.traceId,
            result.errors.join("; ")
          )
        );
      }

      return result;
    } catch (err) {
      const durationMs = performance.now() - start;
      const errorMsg = err instanceof Error ? err.message : String(err);

      logger.emit(
        agentFailedEvent(agentId, context.tenantId, context.traceId, errorMsg)
      );

      return agentResultFail(context, [errorMsg], durationMs);
    }
  }

  /**
   * Mid-execution policy check for a specific action.
   * Agents should call this before performing sensitive operations.
   */
  checkAction(
    context: AgentContext,
    action: string,
    environment?: string
  ): PolicyDecision {
    const agentId = getAgentId(context);
    const logger = new AuditLogger(context.tenantId, agentId, context.traceId);

    const decision = this._policy.evaluate(
      context.dsl,
      action,
      environment ?? context.environment
    );

    if (decision.allowed) {
      logger.logRaw(AuditEventType.POLICY_ALLOWED, { action });
    } else {
      logger.emit(
        policyBlockedEvent(
          agentId,
          context.tenantId,
          context.traceId,
          action,
          decision.reason
        )
      );
    }

    return decision;
  }
}
